hibernate-core/src/main/java/org/hibernate/cfg/IndexOrUniqueKeySecondPass.java

```diff
@@ -72,7 +72,7 @@ public void doSecondPass(Map persistentClasses) throws MappingException {
 		if ( column != null ) {
 			this.table = column.getTable();
 
-			PersistentClass persistentClass = (PersistentClass) persistentClasses.get( column.getPropertyHolder().getEntityName() );
+			PersistentClass persistentClass = (PersistentClass) persistentClasses.get( column.getPropertyHolder().getEntityOwnerClassName() );
 			Property property = persistentClass.getProperty( column.getPropertyName() );
```

Comment:
Effectively, my solution seems buggy. It is surprising that it does not trigger any test failure.

Maybe adding an indexed field on your User entity (with its custom entity name) would be a good move to prevent regression on this use-case.